File: collie/src/main/java/com/snail/collie/core/LooperMonitor.java
package com.snail.collie.core;

import android.os.Build;
import android.os.Looper;
import android.os.MessageQueue;
import android.os.SystemClock;
import android.util.Log;
import android.util.Printer;

import androidx.annotation.CallSuper;


import java.util.HashSet;
import java.util.Objects;

public class LooperMonitor implements MessageQueue.IdleHandler {

    private final HashSet<LooperDispatchListener> listeners = new HashSet<>();
    private static final String TAG = "LooperMonitor";
    private LooperPrinter printer;
    private Looper looper;
    private static final long CHECK_TIME = 60 * 1000L;
    private long lastCheckPrinterTime = 0;

    public abstract static class LooperDispatchListener {

        boolean isHasDispatchStart = false;

        public boolean isValid() {
            return false;
        }


        public void dispatchStart() {

        }

        @CallSuper
        public void onDispatchStart(String x) {
            this.isHasDispatchStart = true;
            dispatchStart();
        }

        @CallSuper
        public void onDispatchEnd(String x) {
            this.isHasDispatchStart = false;
            dispatchEnd();
        }


        public void dispatchEnd() {
        }
    }

    private static final LooperMonitor mainMonitor = new LooperMonitor();

    public static void register(LooperDispatchListener listener) {
        mainMonitor.addListener(listener);
    }

    public static void unregister(LooperDispatchListener listener) {
        mainMonitor.removeListener(listener);
    }

    public HashSet<LooperDispatchListener> getListeners() {
        return listeners;
    }

    public void addListener(LooperDispatchListener listener) {
        synchronized (listeners) {
            listeners.add(listener);
        }
    }

    public void removeListener(LooperDispatchListener listener) {
        synchronized (listeners) {
            listeners.remove(listener);
        }
    }

    public LooperMonitor(Looper looper) {
        Objects.requireNonNull(looper);
        this.looper = looper;
        resetPrinter();
        addIdleHandler(looper);
    }

    private LooperMonitor() {
        this(Looper.getMainLooper());
    }

    @Override
    public boolean queueIdle() {
        if (SystemClock.uptimeMillis() - lastCheckPrinterTime >= CHECK_TIME) {
            resetPrinter();
            lastCheckPrinterTime = SystemClock.uptimeMillis();
        }
        return true;
    }

    public synchronized void release() {
        if (printer != null) {
            synchronized (listeners) {
                listeners.clear();
            }
            looper.setMessageLogging(printer.origin);
            removeIdleHandler(looper);
            looper = null;
            printer = null;
        }
    }

    private static boolean isReflectLoggingError = false;

    private synchronized void resetPrinter() {
        Printer originPrinter = null;
        try {
            if (!isReflectLoggingError) {
                originPrinter = ReflectUtils.get(looper.getClass(), "mLogging", looper);
                if (originPrinter == printer && null != printer) {
                    return;
                }
            }
        } catch (Exception e) {
            isReflectLoggingError = true;
        }

        looper.setMessageLogging(printer = new LooperPrinter(originPrinter));
    }

    private synchronized void removeIdleHandler(Looper looper) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            looper.getQueue().removeIdleHandler(this);
        } else {
            try {
                MessageQueue queue = ReflectUtils.get(looper.getClass(), "mQueue", looper);
                queue.removeIdleHandler(this);
            } catch (Exception e) {
                Log.e(TAG, "[removeIdleHandler] %s", e);
            }

        }
    }

    private synchronized void addIdleHandler(Looper looper) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            looper.getQueue().addIdleHandler(this);
        } else {
            try {
                MessageQueue queue = ReflectUtils.get(looper.getClass(), "mQueue", looper);
                queue.addIdleHandler(this);
            } catch (Exception e) {
                Log.e(TAG, "[removeIdleHandler] %s", e);
            }
        }
    }


    class LooperPrinter implements Printer {
        public Printer origin;
        boolean isHasChecked = false;
        boolean isValid = false;

        LooperPrinter(Printer printer) {
            this.origin = printer;
        }

        @Override
        public void println(String x) {
            if (null != origin) {
                origin.println(x);
                if (origin == this) {
                    throw new RuntimeException(TAG + " origin == this");
                }
            }

            if (!isHasChecked) {
                isValid = x.charAt(0) == '>' || x.charAt(0) == '<';
                isHasChecked = true;

            }

            if (isValid) {
                dispatch(x.charAt(0) == '>', x);
            }

        }
    }


    private void dispatch(boolean isBegin, String log) {

        for (LooperDispatchListener listener : listeners) {
            if (listener.isValid()) {
                if (isBegin) {
                    if (!listener.isHasDispatchStart) {
                        listener.onDispatchStart(log);
                    }
                } else {
                    if (listener.isHasDispatchStart) {
                        listener.onDispatchEnd(log);
                    }
                }
            } else if (!isBegin && listener.isHasDispatchStart) {
                listener.dispatchEnd();
            }
        }

    }


}

File: collie/src/main/java/com/snail/collie/core/LooperMonitor.kt
package com.snail.collie.core

import android.os.Build
import android.os.Looper
import android.os.MessageQueue
import android.os.SystemClock
import android.util.Printer
import androidx.annotation.CallSuper
import java.lang.Exception
import java.util.*

object LooperMonitor : MessageQueue.IdleHandler {

    private var looper: Looper
    private var printer: LooperPrinter? = null

    init {
        Objects.requireNonNull(Looper.getMainLooper())
        looper = Looper.getMainLooper()
        resetPrinter()
        addIdleHandler(looper)
    }

    @Synchronized
    private fun addIdleHandler(looper: Looper) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            looper.queue.addIdleHandler(this)
        } else {
            try {
                val queue = ReflectUtils.get<MessageQueue>(looper?.javaClass, "mQueue", looper)
                queue?.addIdleHandler(this)
            } catch (e: Exception) {
            }
        }
    }


    @Synchronized
    private fun removeIdleHandler(looper: Looper) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            looper.queue.removeIdleHandler(this)
        } else {
            try {
                val queue = ReflectUtils.get<MessageQueue>(looper.javaClass, "mQueue", looper)
                queue?.removeIdleHandler(this)
            } catch (e: Exception) {
            }
        }
    }

    private const val CHECK_TIME = 60 * 1000L
    private var lastCheckPrinterTime: Long = 0

    override fun queueIdle(): Boolean {
        //  定期重置
        if (SystemClock.uptimeMillis() - lastCheckPrinterTime >= LooperMonitor.CHECK_TIME) {
            resetPrinter()
            lastCheckPrinterTime = SystemClock.uptimeMillis()
        }
        return true
    }

    private var isReflectLoggingError = false

    @Synchronized
    private fun resetPrinter() {
        var originPrinter: Printer? = null
        try {
            if (!isReflectLoggingError) {
                originPrinter = ReflectUtils.get(looper.javaClass, "mLogging", looper)
                if (originPrinter === printer && null != printer) {
                    return
                }
            }
        } catch (e: Exception) {
            isReflectLoggingError = true
        }
        looper.setMessageLogging(LooperPrinter(originPrinter).also { printer = it })
    }

    class LooperPrinter(var originPrinter: Printer?) : Printer {

        override fun println(x: String?) {

            originPrinter?.let {
                if (it == this@LooperPrinter) {
                    return
                }
                it.println(x)
            }
            x?.let {
                if (x[0] == '>' || x[0] == '<') {
                    dispatch(x[0] == '>', x)
                }
            }
        }
    }

    abstract class LooperDispatchListener {
        var isHasDispatchStart = false

        open fun dispatchStart() {}
        open fun dispatchEnd() {}

        @CallSuper
        fun onDispatchStart(x: String?) {
            isHasDispatchStart = true
            dispatchStart()
        }

        @CallSuper
        fun onDispatchEnd(x: String?) {
            isHasDispatchStart = false
            dispatchEnd()
        }
    }

    private val listeners = mutableSetOf<LooperDispatchListener>()

    private fun dispatch(isBegin: Boolean, log: String) {
        for (listener in listeners) {
            if (isBegin) {
                if (!listener.isHasDispatchStart) {
                    listener.onDispatchStart(log)
                }
            } else {
                if (listener.isHasDispatchStart) {
                    listener.onDispatchEnd(log)
                }
            }
        }
    }

    @Synchronized
    fun release() {
        if (printer != null) {
            listeners.clear()
            looper.setMessageLogging(printer?.originPrinter)
            removeIdleHandler(looper)
            printer = null
        }
    }

    fun register(listener: LooperDispatchListener) {
        listeners.add(listener)
    }

    fun unregister(listener: LooperDispatchListener?) {
        listeners.remove(listener)
    }
}
